pkg/ivd/ivd_protected_entity_type_manager_util.go

```diff
@@ -0,0 +1,204 @@
+package ivd
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/cns"
+	cnstypes "github.com/vmware/govmomi/cns/types"
+	"github.com/vmware/govmomi/find"
+	"github.com/vmware/govmomi/vim25"
+	vim25types "github.com/vmware/govmomi/vim25/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"strings"
+)
+
+func findAllDatastores(ctx context.Context, client *vim25.Client) ([]vim25types.ManagedObjectReference, error) {
+	finder := find.NewFinder(client)
+	dss, err := finder.DatastoreList(ctx, "*")
+	if err != nil {
+		return nil, err
+	}
+
+	var dsList []vim25types.ManagedObjectReference
+	for _, ds := range dss {
+		dsList = append(dsList, ds.Reference())
+	}
+	if len(dsList) == 0 {
+		return nil, errors.New("No datastore can be found")
+	}
+
+	return dsList, nil
+}
+
+func retrievePlatformInfoFromConfig(config *rest.Config, params map[string]interface{}, logger logrus.FieldLogger) error {
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to get k8s clientset from the given config: %v", config)
+		return err
+	}
+
+	ns := "kube-system"
+	secretApis := clientset.CoreV1().Secrets(ns)
+	vsphere_secret := "vsphere-config-secret"
+	secret, err := secretApis.Get(vsphere_secret, metav1.GetOptions{})
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to get k8s secret, %s", vsphere_secret)
+		return err
+	}
+	sEnc := string(secret.Data["csi-vsphere.conf"])
+	lines := strings.Split(sEnc, "\n")
+
+	for _, line := range lines {
+		if strings.Contains(line, "VirtualCenter") {
+			parts := strings.Split(line, "\"")
+			params["VirtualCenter"] = parts[1]
+		} else if strings.Contains(line, "=") {
+			parts := strings.Split(line, "=")
+			key := strings.TrimSpace(parts[0])
+			value := strings.TrimSpace(parts[1])
+			params[key] = value[1 : len(value)-1]
+		}
+	}
+
+	return nil
+}
+
+func createCnsVolumeWithClusterConfig(ctx context.Context, config *rest.Config, client *govmomi.Client, md metadata, logger logrus.FieldLogger) (string, error) {
+	params := make(map[string]interface{})
+	err := retrievePlatformInfoFromConfig(config, params, logger)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to retrieve VC config secret: %+v", err)
+		return "", err
+	}
+
+	clusterId, ok := params["cluster-id"].(string)
+	if !ok {
+		logger.WithError(err).Errorf("Failed to retrieve cluster id")
+		return "", err
+	}
+
+	user, ok := params["user"].(string)
+	if !ok {
+		logger.WithError(err).Errorf("Failed to retrieve vsphere user")
+		return "", err
+	}
+	logger.Debugf("Retrieved cluster id, %v, and vSphere user, %v", clusterId, user)
+
+	// Preparing for the VolumeCreateSpec for the volume provisioning
+	logger.Debug("Preparing for the VolumeCreateSpec for the volume provisioning")
+	dsList, err := findAllDatastores(ctx, client.Client)
```

Comment:
Is this right?  Are we going to pass ALL datastores into the create spec?  What happens when we have 2K datastores?

Reply:
First of all, they are just in-memory operations. Secondly, CNS API will pick up the datastore based on the storage policy and the maximum available space. no datastore traversal is involved. So, even in the case with 2K datastores, it should not introduce much overhead.